### PR TITLE
Debounce selector

### PR DESCRIPTION
**Pull request by frankpepermans**
Small change for the `DebounceTransformer`, which now accepts a selector instead of a fixed `Duration`.

As requested in https://github.com/ReactiveX/rxdart/issues/168

On our `Observable`, we now have:
```
debounce(Duration duration);
debounceSelector(Duration durationSelector(T event));
```

### CODE REVIEW

**Comment by brianegan** [comment]
One quick Q: Looking at the RxJS implementation, the durationSelector returns an Observable, rather than a `Duration`. Think we should do the same?

**Comment by frankpepermans** [comment]
Oh then we'd need schedulers, like the rxjs timer. Maybe a different name for this one? I.e. debounceFor, debounceWith, ...

Or simply debounce with 2 parameters, so you choose to provide either a handler or a fixed Duration?

**Comment by frankpepermans** [comment]
Looking at https://github.com/ReactiveX/rxjs/blob/master/src/internal/operators/debounce.ts, it accepts either an `Observable` or a `Promise` only. In fact all time-based implementations in rxjs do this, i.e. https://github.com/ReactiveX/rxjs/blob/master/src/internal/operators/throttle.ts as well.

So maybe we oughta consider changing this *everywhere* at some point.

Now for this PR, in the same context it might be better to just cancel it at this point, this PR would create another `debounceX` next to the old `debounce`, but we should be consistent then and do the same for `throttle` and `sample` for example.

If we then at some point change the API to accommodate to rxjs, we'd then have to drop these extra ops again.

So TL;DR we delete this PR and match rxjs better in the Future, for all timer ops?
